plugins/rest/aws.go

```diff
@@ -132,27 +151,31 @@ func (cs *awsMetadataCredentialService) refreshFromService() error {
 		return err
 	}
 
-	resp, err := http.Get(metaDataURL)
+	// construct an HTTP client with a reasonably short timeout
+	client := &http.Client{Timeout: time.Second * 10}
+	req, err := http.NewRequest(http.MethodGet, metaDataURL, nil)
 	if err != nil {
-		// some kind of catastrophe talking to the EC2 metadata service
-		return err
+		return errors.New("unable to construct metadata HTTP request: " + err.Error())
 	}
-	defer resp.Body.Close()
 
-	logrus.WithFields(logrus.Fields{
-		"url":     metaDataURL,
-		"status":  resp.Status,
-		"headers": resp.Header,
-	}).Debug("Received response from metadata service.")
-
-	if resp.StatusCode != 200 {
-		// most probably a 404 due to a role that's not available; but cover all the bases
-		return errors.New("metadata service HTTP request failed: " + resp.Status)
+	// if in the EC2 environment, we will use IMDSv2, which requires a session cookie from a
+	// PUT request on the token endpoint before it will give the credentials, this provides
+	// protection from SSRF attacks
+	if !isECS() {
```

Comment:
I'm not super familiar with the service, so maybe a dumb question, but can't we use this IMDSv2 stuff even if it is in ECS? Or is that metadata service not available there?

Reply:
It does not seem to be available and I can see why Amazon thought it was not necessary - rather than having a well-known fixed endpoint for the metadata service (as in EC2), the path to the equivalent ECS endpoint has an embedded UUID that's passed into the container as an environment variable, `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` ends up with a value like `/v2/credentials/be3d4d7c-28b6-47e6-8081-cd746d95cb28`. 

So an attacker has to guess a 128 bit random value or be able to force the container to disclose its environment to mount a similar attack.

Reply:
I should add; I did cross-check this against the botocore implementation which I'm taking to be canonical, and it's consistent.

ref. https://github.com/boto/botocore/blob/f1f0c6d2e445d7c3a3f8f355eaf6d692bbf3cd6a/botocore/utils.py#L1582 et seq

Reply:
Good to know! Thank you for the detailed explanation 👍